sysl2/sysl/ints.go

```diff
@@ -0,0 +1,219 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type IntegrationParam struct {
+	m            *sysl.Module
+	excludes     StrSet
+	passthroughs StrSet
+	commonEndpts StrSet
+	highlights   StrSet
+	apps         StrSet
+	deps         map[string]*AppDependency
+}
+
+func MakeIntegrationParam(mod *sysl.Module, excludes, passthroughs, highlights StrSet) *IntegrationParam {
+	return &IntegrationParam{
+		m:            mod,
+		excludes:     excludes,
+		passthroughs: passthroughs,
+		commonEndpts: MakeStrSet(".. * <- *", "*"),
+		highlights:   highlights,
+		apps:         StrSet{},
+		deps:         map[string]*AppDependency{},
+	}
+}
+
+func (v *IntegrationParam) collectApplicationDependencies(apps StrSet) {
+	for app := range apps {
+		for epname, endpt := range v.m.GetApps()[app].GetEndpoints() {
+			v.handleStatement(app, epname, endpt.GetStmt())
+		}
+	}
+}
+
+func (v *IntegrationParam) handleStatement(appname, epname string, stmts []*sysl.Statement) {
+	for _, stmt := range stmts {
+		switch t := stmt.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			targetName := getAppName(t.Call.GetTarget())
+			if ok := v.addAppDependency(appname, epname, targetName, t.Call.Endpoint, stmt); !ok {
+				continue
+			}
+			v.addEndpoint(targetName, t.Call.Endpoint)
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			v.handleStatement(appname, epname, t.Cond.GetStmt())
+		case *sysl.Statement_Loop:
+			v.handleStatement(appname, epname, t.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			v.handleStatement(appname, epname, t.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			v.handleStatement(appname, epname, t.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			v.handleStatement(appname, epname, t.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range t.Alt.GetChoice() {
+				v.handleStatement(appname, epname, choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func (v *IntegrationParam) addAppDependency(sourceApp, sourceEndpt, targetApp, targetEndpt string, stmt *sysl.Statement) bool {
+	if len(v.apps) == 0 {
+		v.apps = v.apps.Union(v.highlights)
+	}
+	apps := MakeStrSet(sourceApp, targetApp)
+	if len(apps.Intersection(v.excludes)) > 0 {
+		return false
+	}
+	if len(apps.Intersection(v.highlights)) > 0 {
+		if !hasPattern("human", v.m.GetApps()[sourceApp].GetAttrs()) {
+			v.apps.Insert(sourceApp)
+		}
+		if !hasPattern("human", v.m.GetApps()[targetApp].GetAttrs()) {
+			v.apps.Insert(targetApp)
+		}
+	}
+	if !(apps.IsSubSet(v.apps) || (v.apps.Contains(sourceApp) && v.passthroughs.Contains(targetApp)) || v.passthroughs.Contains(sourceApp)) {
+		return false
+	}
+	if len(MakeStrSet(sourceEndpt, targetEndpt).Intersection(v.commonEndpts)) > 0 {
```

Comment:
again unncessary use of set.
v. commonEndpts has only one element EVER. i.e. `".. * <- *", "*"`.
and we will never get this case, because you dont call the collection endpoint.
But if you need this check, its probably the 1st check you should do in this function.
```go
if sourceEndpt == ".. * <- *", "*" || targetEndpt == ".. * <- *", "*" {
  return false
}
```

Reply:
moved

Reply:
but there are actually two elements` ".. * <- *" and "*"`